ideafast/middleware-services
# Implement test automation (nox) and linting and typing through pre-commit. Resolves #22

This PR implements linting, typing and tests such that upcoming contributions are sound - this should help with error prevention in the long run. 

More specifically, changes include (as suggested in #22):
- set up `nox` for test automation
- set up `nox` session to run:
   - `flake8` linting
   - `pytests` unit testing
   - `mypy` type testing
   - `isort` import statement ordering
   - `black` automatic code formatting
- set up `pre-commit` to run (and mandate passes) for these, except `pytests`. 

As expected, these prompted a breadth of required changes which are all resolved. A few simply cannot due to our setup, and have been added to config files for `mypy` and `flake8` to be ignored.

## Test
- pull down this branch
- run `poetry install` to include dev dependencies
- enter a virtual environment (`poetry shell`)
- set up pre-commit hooks: `pre-commit install`
- run individual check:
   - `poetry run nox -rs tests`. This runs all tests (and currently fails a few on the inventory part). Note that it currently circumvents `install_with_constraints_. See issue #36.
   - `poetry run nox -rs mypy`. This checks typing. Note that a few tests are currently excluded (see issue #35).
   - `poetry run nox -rs isort`. This sorts import statements alphabetically and by sections (i.e. global, 3rd party and local)
   - `poetry run nox -rs lint`. This runs various linting checks using flake8. Note that a few errors are ignored for our use case - see _mypy.ini_ in root.
   - `poetry run nox -rs black`. This checks **and changes** various formatting following _black_ standards.
- to test pre-commit functionality, make a simple change and commit. This should run (and pass) 
   - Check Yaml
   - Trim Trailing Whitespace
   - black
   - flake8
   - isort, and
   - mypy

## To Do
As referred to above, a few issues and todo's arose from the adoption of linting, tests and typing. These are documented in separate issues to work on.
